flexprep/domain/processing.py

```diff
@@ -122,20 +121,40 @@ def _save_output(
         forecast_ref_time_str = forecast_ref_time.strftime("%Y%m%d%H%M")
 
         try:
-            key = f"output_dispf{forecast_ref_time_str}_step{step_to_process}"
+            key = f"output_dispf{forecast_ref_time_str}{step_to_process}"
+
+            ref = next(
+                f
+                for f in ds_out.values()
+                if metadata.extract_keys(f.message, "editionNumber") == 2
+            )
 
             with tempfile.NamedTemporaryFile(
-                suffix=key,
+                suffix=key, delete=False, mode="wb"
             ) as output_file:
-                # Write data to the temporary file
-                with open(output_file.name, "wb") as fout:
-                    for name, field in ds_out.items():
-                        if field.attrs.get("v_coord") == "hybrid":
-                            logger.info(f"Writing GRIB fields to {output_file.name}")
-                            grib_decoder.save(field, fout)
-
-                    # Upload the file to S3
-                    S3client().upload_file(output_file.name, key=key)
+                for name, field in ds_out.items():
+                    if field.squeeze().isnull().all():
+                        logging.info(f"Ignoring field {field} - only NaN values")
+                        continue
+
+                    if metadata.extract_keys(field.message, "editionNumber") == 1:
+                        if name in ["lsp", "sshf", "ewss", "nsss"]:
+                            msg = metadata.override(
+                                ref.message,
+                                productDefinitionTemplateNumber=8,
+                                shortName=field.parameter["shortName"],
+                            )
+                        else:
+                            msg = metadata.override(
+                                ref.message, shortName=field.parameter["shortName"]
+                            )
+                        field.attrs = msg
+                    grib_decoder.save(field, output_file)
+            logger.info("Writing GRIB fields to file completed.")
+
+            # Upload the file to S3
+            S3client().upload_file(output_file.name, key=key)
```

Comment:
Why did you move the upload out of the context manager ?

Reply:
I moved it outside the context manager to ensure the file is fully written and closed before proceeding with the upload. Does it make more sense for you to leave it inside?

Reply:
File operations are blocking by default, meaning that the file contents are updated (no longer in the buffer) when the call returns. The purpose of the context manager is to remove the file if an exception is encountered. In this case, it makes sense to me to leave the upload inside the context manager.